pkg/pillar/hardware/smart.go

```diff
@@ -0,0 +1,227 @@
+	smart.PowerCycleCount = getAttrFromJSON(attrList, idPowerCycleCount)
+	smart.PowerOnHours = getAttrFromJSON(attrList, idPowerOnHours)
+	smart.ReallocatedSectorCt = getAttrFromJSON(attrList, idReallocatedSectorCt)
+	return smart
+}
+
+// ReadDiskSmartInfoFromJSON read JSON file and return *types.DeviceSmartInfo
+func ReadDiskSmartInfoFromJSON(diskName string) (*types.DeviceSmartInfo, error) {
+	path := getPathForFileDisk(diskName)
+	data, err := readFWithMaxSize(path, maxSmartCtlSize)
+	if err != nil {
+		return nil, fmt.Errorf("read file with SMART info for %s disk failed. %s", path, err)
+	}
+
+	deviceInfo := new(types.DeviceSmartInfo)
+	if err := json.Unmarshal(data, &deviceInfo); err != nil {
+		return nil, fmt.Errorf("error while parsing SMART data. %s", err)
+	}
+
+	if deviceInfo.Smartctl.ExitStatus != 0 {
+		return nil, fmt.Errorf("error-code: %d, info:%v",
+			deviceInfo.Smartctl.ExitStatus,
+			deviceInfo.Smartctl.Messages)
+	}
+
+	return deviceInfo, nil
+}
+
+// getAndComparisonSmartDataForDisk takes a disk name as input, then tries
+// to collect old data and get new data, compares if necessary and
+// always returns new data in the form of:
+//
+// *types.DeviceSmartInfo - regardless of the result, we always send last new data,
+// except for situations with errors when receiving this data
+//
+// bool - Need to update information, true - necessary, false - data is the same
+//
+// error - Errors with getting new data, if any
+func getAndComparisonSmartDataForDisk(diskName string) (*types.DeviceSmartInfo, bool, error) {
+	needUpdate := false
+	oldData, err := ReadDiskSmartInfoFromJSON(diskName)
+	if err != nil {
+		// it's normal situation
+		needUpdate = true
+	}
+
+	// Get new S.M.A.R.T data for disks
+	args := []string{"--all", diskName, "--json"}
+	pathFile := getPathForFileDisk(diskName)
+	if err := smartctlExec(pathFile, args...); err != nil {
+		return nil, false, fmt.Errorf("get disk info from smartctl filed. err: %s", err)
+	}
+
+	newData, err := ReadDiskSmartInfoFromJSON(diskName)
+	if err != nil {
+		return nil, false, fmt.Errorf("new SMART data for disk %s not found", diskName)
+	}
+
+	if needUpdate {
+		return newData, needUpdate, nil // no have old info, need send new info
+	}
+
+	if reflect.DeepEqual(oldData, newData) {
+		needUpdate = false // the data is the same
+	} else {
+		needUpdate = true
+	}
+
+	// Regardless of the result, we always send last new data
+	return newData, needUpdate, nil
+}
+
+// GetStorageDiskInfo takes a disk name as input, returns info.StorageDiskInfo
+func GetStorageDiskInfo(diskName string) (*info.StorageDiskInfo, error) {
+	stDiskInfo := new(info.StorageDiskInfo)
+	deviceInfo, _, err := getAndComparisonSmartDataForDisk(diskName)
+	if err != nil {
+		return nil, err
+	}
+
+	//if needUpdate {
+	//	Here, if needed in the future, you can implement
+	//	logic with a reaction to data changes.
+	//	But now the data will be sent anyway.
+	//}
```

Comment:
I suspect we need logic to check for changes in zedagent so that it will trigger a device info message when there is a change to the smartdata.